Improve the state.js contract error handling logic (Closes #1165)

taoeffect opened the pull request:
### Changes

+ Moved all Chelonia-related Vuex stuff into Chelonia SBP framework (Closes #1165)
+ Autoban after contract sync (Closes #731)
+ Removed Mailbox.vue (Closes #1217)
+ May have fixed blank white page on large amount of events (#1192)
+ Added debounce to contract sync banner (Closes #1220)
+ Migrate to NPM SBP stuff (Closes #1153 - my mistake for not assigning to myself when @SebinSong did the same thing!)
+ Added a-link support to `v-safe-html` via `v-safe-html:a`
+ Moved `logExceptNavigationDuplicated` into `@view-utils` since it's a vue-related utility
+ Removed TimeTravel.vue
+ Added some more hacks for esbuild failing to load chunks for chatroom.js contract
+ Changed pubsub logging from `.log` to `.debug` and `.warn` and `.error` as necessary
+ I don't know if this PR closes #1176 or not, but the error is gone. However, u3 is now able to rejoin the group after they've been kicked using the original invite link. Should open a new issue for that
+ Display user notifications on contract errors
+ Updated `esbuild` to `0.14.38` and `esbuild-sass-plugin` to `2.2.6` 

The new contract error handling rules are:

- Exceptions thrown during mutation processing (e.g. either by `validate` or by `process` functions) will result in that message being ignored and any side-effects skipped. Unless `ChelErrorUnrecoverable` is thrown, in which case that contract comes to a permanent halt (usually not something you want).
  - Until #1214 is closed, this rule can lead to divergent state among clients. Example: a process function throws an exception and is ignored. Later that contract is updated to not throw the exception. Clients who join later will process the mutation, resulting in a different state from older clients.
- Exceptions thrown during side-effect processing (e.g. either by a pushed side effect or in the `sideEffect` function or in the `postHandleEvent` hook or in an event listener) will result in that message getting dropped and the application coming to a halt until the bug in the side effect is fixed, after which a page refresh will resume normal functioning of the app.

Additionally, applications can modify the behavior of exceptions of process functions by implementing the `processError` hook. For example, if `group.js` throws `GIErrorIgnoreAndBan`, it will autoban that user. Chelonia can be configured with various hooks and event listeners for further customization, see changes in `main.js` for more examples.

taoeffect committed:
beginnings of refactoring to move Vuex logic into chelonia

taoeffect committed:
added @sbp/sbp (Closes #1153)

taoeffect committed:
bunch of stuff

taoeffect committed:
added some basic error handling

taoeffect committed:
autoban only after contract sync (Closes #731)

taoeffect committed:
fix flow errors

taoeffect committed:
added a link support to v-safe-html in special cases + fixed some tests and errors

taoeffect committed:
removed Mailbox stuff (Closes #1217) + removed some vuex mutations + added skipSideEffects

taoeffect committed:
improved backend.test.js a bit

taoeffect committed:
remember to set reactiveSet and reactiveDel

snowteamer commented on backend/database.js:
```diff
@@ -1,13 +1,13 @@
-import '~/shared/domains/okTurtles/data.js'
-import '~/shared/domains/gi/db.js'
+import '@sbp/okturtles.data'
+import '~/shared/domains/chelonia/db.js'
```
So the DB stuff is now a part of Chelonia?

taoeffect replied:
Yep. Some selectors are designed to be overwritten by the app

snowteamer commented on backend/routes.js:
```diff
@@ -92,7 +92,7 @@ route.GET('/name/{name}', {}, async function (request, h) {
 route.GET('/latestHash/{contractID}', {}, async function (request, h) {
   try {
-    const entry = await sbp('gi.db/log/lastEntry', request.params.contractID)
+    const entry = await sbp('chelonia/db/lastEntry', request.params.contractID)
```
Shouldn't there be a `'chelonia/db/latestHash'` selector? This way a route handler wouldn't have to know it needs to call `.hash()` on an entry to get its hash.

snowteamer commented on frontend/controller/actions/group.js:
```diff
@@ -345,6 +348,80 @@ export default (sbp('sbp/selectors/register', {
+            proposalHash = hash
+            break
+          }
+        }
```
This could be made more concise, for example:
```suggestion
        // Find existing proposal if it exists.
        let [ proposalHash, proposal ] = Object.entries(contractState.proposals)
          .find(([hash, prop]) => (
            prop.status === STATUS_OPEN &&
            prop.data.proposalType === PROPOSAL_REMOVE_MEMBER &&
            prop.data.proposalData.member === username
          )) ?? []
```

taoeffect replied:
Yeah that looks like a good improvement, but is it me or would this code throw some kind of an error?

```
> d = { a: 4 }
{ a: 4 }
> Object.entries(d)
[ [ 'a', 4 ] ]
> let [i, j] = Object.entries(d).find(() => false)
Uncaught:
TypeError: undefined is not iterable (cannot read property Symbol(Symbol.iterator))
```

Or am I doing something wrong?

Silver-IT replied:
> let [i, j] = Object.entries(d).find(() => false)

Two variables `i` and `i` are set from the result of `find` which returns `undefined` always.
So it's just like `let [i, j] = undefined;`

If you try with the following code, it won't make any errors if object `d` is not empty.
`let [i, j] = Object.entries(d).find(() => true)`

taoeffect replied:
Ah, but that's the thing, `proposals` might not contain the proposal we're looking for, so in that case this would error.

snowteamer replied:
Sorry for leaving this case out.

Update: appended `?? []` to the suggested code to make it return and assign `undefined` values in case no proposal is found. But I find it a bit less clean now.

snowteamer commented on frontend/controller/backend.js:
```diff
@@ -2,81 +2,18 @@
-})
-
-sbp('okTurtles.events/on', GI_UPDATE_AVAILABLE, (version) => {
+sbp('okTurtles.events/on', NOTIFICATION_TYPE.APP_VERSION, (version) => {
```
Strictly speaking, `APP_VERSION` is a message type not an event type, but I guess that's an acceptable shortcut.

taoeffect replied:
Good point. I updated this since we're listening for a framework-provided (not GI-provided) event

snowteamer commented on frontend/views/components/banners/BannerScoped.vue:
```diff
@@ -27,6 +28,10 @@ export default ({
       type: String,
       default: 'feedbackMsg'
+    },
+    allowA: {
```
What does `allowA` mean?

taoeffect replied:
It means allow `a` tags to be rendered

snowteamer commented on frontend/views/containers/chatroom/ChatMain.vue:
```diff
@@ -83,7 +83,7 @@ import { MESSAGE_TYPES, MESSAGE_ACTION_TYPES, MESSAGE_VARIANTS } from '@model/co
 import { proximityDate, MINS_MILLIS } from '@utils/time.js'
 import { CHATROOM_MESSAGE_ACTION, CHATROOM_STATE_LOADED } from '~/frontend/utils/events.js'
-import { CONTRACT_IS_SYNCING } from '@utils/events.js'
+import { CONTRACT_IS_SYNCING } from '~/shared/domains/chelonia/chelonia.js'
```
No `'~/shared/domains/chelonia/events.js'` file from where to import event names?

taoeffect replied:
Yeah that would be nicer, good suggestion

snowteamer commented on frontend/views/utils/vSafeHtml.js:
```diff
@@ -29,8 +30,12 @@ export const defaultConfig = {
+      config = cloneDeep(config)
+      config.ALLOWED_ATTR.splice(config.ALLOWED_ATTR.length, 0, 'href', 'target')
+      config.ALLOWED_TAGS.push('a')
+    }
```
Hmm now I understand that `allowA` was meant for allowing anchor tags in `v-safe-html `directives.
Now that I remember, there is a configurable safe URL whitelist in `utils/allowedUrls.js`, couldn't it have been used?

snowteamer commented on shared/domains/chelonia/GIMessage.js:
```diff
@@ -46,8 +46,7 @@ export class GIMessage {
-    const instance = new this()
-    instance._message = {
+    const message = {
       version: 1.00,
```
Shouldn't `version` be a SemVer string instead?

taoeffect committed:
update esbuild and remove TimeTravel.vue + added more hacks for chunks

taoeffect committed:
fixes several Cypress tests

taoeffect committed:
added and used 'chelonia/pubsub/update' upon login

taoeffect committed:
fixed display of notifications on contract errors + may have fixed blank white page (closes #1192)

snowteamer commented on shared/domains/chelonia/chelonia.js:
```diff
@@ -62,18 +91,60 @@ sbp('sbp/selectors/register', {
+  'chelonia/connect': function (): Object {
+    if (!this.config.connectionURL) throw new Error('config.connectionURL missing')
+    if (!this.config.connectionOptions) throw new Error('config.connectionOptions missing')
+    if (this.socket) {
```
Since `this.socket` here is actually not a WebSocket object, I think something like `this.client` or `this.pubsub` would be more correct

taoeffect committed:
@snowteamer PR review + debounce the banner on contract sync (Closes #1220)

taoeffect committed:
fixed all Cypress tests, verified

taoeffect committed:
remove 2nd esbuild hack

taoeffect committed:
update Information-Flow.md + updated & fixed strings

snowteamer commented on shared/pubsub.js:
```diff
@@ -243,7 +244,7 @@ const defaultClientEventHandlers = {
     const client = this
 
-    console.log('[pubsub] Event: error', event)
+    console.error('[pubsub] Event: error', event)
```
Not all `error` events should be logged with `console.error`, for example every failed connection attempt generates one such event.

taoeffect replied:
noted! will be changed to `console.warn`

snowteamer commented on docs/Information-Flow.md:
```diff
@@ -70,19 +70,19 @@ In our example above, the name of the *contract action* generated by `sbp('gi.ac
+// OR wait until it finishes syncing:
 
-await sbp('gi.actions/contract/syncAndWait', contractId)
+await sbp('chelonia/contract/sync', contractID)
```
Nice API :)

taoeffect replied:
Thanks :)

snowteamer commented on frontend/main.js:
```diff
@@ -127,21 +146,41 @@ async function startApp () {
+          bannerGeneral.show(L('Loading events from server...'), 'wifi')
+          clearBannerTimer = setTimeout(clearBanner, 2000)
+        }
+      }, 2000)
```
I was working on that issue too. Checking whether the banner has already a higher severity status is a great idea!

For my part, my solution filters out contract IDs from `state.pending`, to avoid triggering such banner activity when signing up or creating a new group.

snowteamer replied:
In the related issue however, the delays were different, like "show the banner after 1s then don't hide it again before another 2s"

snowteamer replied:
Also I feel like more of this banner-related code should be moved to `BannerGeneral.vue` rather than stay in `main.js`.

taoeffect replied:
> In the related issue however, the delays were different, like "show the banner after 1s then don't hide it again before another 2s"

Yeah, I edited the issue because I decided they should be slightly tweaked. 1 second to read a message seemed a bit short to me.

> Also I feel like more of this banner-related code should be moved to BannerGeneral.vue rather than stay in main.js.

Good point, I'll add a new `debouncedShow` method to BannerGeneral.vue and will push

snowteamer commented on frontend/main.js:
```diff
@@ -194,6 +232,29 @@ async function startApp () {
+      }
+      // useful to delay the initialization of various pages like Join.vue until
+      // we have finished setting up and logging in
+      sbp('okTurtles.events/emit', VUE_LOADED)
```
Awesome idea too!

snowteamer commented on frontend/model/notifications/templates.js:
```diff
@@ -12,6 +12,33 @@ const escapeForHtml = (text) => text.replace(/[<>&]/g, '\\$&')
+      linkTo: `/app/dashboard?modal=UserSettingsModal&section=application-logs&errorMsg=${encodeURI(data.message)}`,
+      scope: 'app'
+    }
+  },
```
At first guess I would say that errors and warnings would rather belong in the global banner or something similar. But maybe it's already used for too many things.

snowteamer commented on frontend/views/utils/vSafeHtml.js:
```diff
@@ -29,8 +30,12 @@ export const defaultConfig = {
+    let config = defaultConfig
+    if (binding.arg === 'a') {
+      config = cloneDeep(config)
+      config.ALLOWED_ATTR.splice(config.ALLOWED_ATTR.length, 0, 'href', 'target')
```
This `splice` call doesn't seem to remove any element, only to add `href` and `target`. So did you mean the following?
```suggestion
      config.ALLOWED_ATTR.push('href', 'target')
```

taoeffect replied:
Lol yes, for some reason I thought `.push` only accepted one argument, learn something new every day. Thanks! :)

taoeffect committed:
@snowteamer review + more:

+ fixed autoban issues
+ added `.debouncedShow` to BannerGeneral.vue
+ allowed BannerGeneral to display links
+ fixed calculation of voting rule for removing member
+ improved strings when a vote has passed or failed
+ fixed some issues with french strings
+ fixed an issue with Cypress failing to run when it can't stringify a circular object

Silver-IT commented on test/cypress/support/commands.js:
```diff
@@ -167,9 +167,6 @@ Cypress.Commands.add('giCreateGroup', (name, {
     cy.getByDT('app').then(([el]) => {
       cy.get(el).should('have.attr', 'data-sync', '')
     })
-
-    checkIfJoinedGeneralChannel(name)
-
```
Actually, I am not sure if this removal breaks cypress test...

taoeffect replied:
They passed though didn't they

Silver-IT replied:
Yeap. Nice catch. It should not be here. It's for ByPassUI.

Silver-IT reviewed (commented):
Actually, I can not say I have understood all the code lines but I agree it's a really impressive and you did a great work on this. Awesome~~~
Just one thing to ask regarding `checkIfJoinedGeneralChannel` function.

Silver-IT reviewed (approved):
Approved. ~~~
